qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,291 @@
+    r"""Perform a non-linear least squares to fit
+
+    This solves the optimization problem
+
+    .. math::
+        \Theta_{\mbox{opt}} = \arg\min_\Theta \sum_i
+            \sigma_i^{-2} (f(x_i, \Theta) -  y_i)^2
+
+    using ``scipy.optimize.curve_fit``.
+
+    Args:
+        func: a fit function `f(x, *params)`.
+        xdata: a 1D float array of x-data.
+        ydata: a 1D float array of y-data.
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        bounds: Optional, lower and upper bounds for optimization
+                parameters.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        result containing `popt` the optimal fit parameters,
+        `popt_err` the standard error estimates popt,
+        `pcov` the covariance matrix for the fit,
+        `chisq` the chi-squared parameter of fit,
+        `xrange` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if the number of y-values in the dataset is not
+                     greater than the number of parameters in the
+                     fit function.
+    """
+    # Format p0 parameters if specified as dictionary
+    if isinstance(p0, dict):
+        param_keys = list(p0.keys())
+        param_p0 = list(p0.values())
+
+        # Convert bounds
+        if bounds:
+            lower = [bounds[key][0] for key in param_keys]
+            upper = [bounds[key][1] for key in param_keys]
+            param_bounds = (lower, upper)
+        else:
+            param_bounds = None
+
+        # Convert fit function
+        def fit_func(x, *params):
+            return func(x, **dict(zip(param_keys, params)))
+
+    else:
+        param_keys = None
+        param_p0 = p0
+        param_bounds = bounds
+        fit_func = func
+
+    # Check data set has at least 1 more values than model parameters
+    num_params = len(param_p0)
+    if len(ydata) <= num_params:
+        raise QiskitError(
+            "The number of y-values in the dataset is not greater than"
+            " the number of parameters in the fit function."
+        )
+
+    # Override scipy.curve_fit default for absolute_sigma to True
+    # if not specified in kwargs
+    if sigma is not None and "absolute_sigma" not in kwargs:
+        kwargs["absolute_sigma"] = True
+
+    # Run curve fit
+    # TODO: Add error handling so if fitting fails we can return an analysis
+    #       result containing this information
+    # pylint: disable = unbalanced-tuple-unpacking
+    popt, pcov = opt.curve_fit(
+        fit_func, xdata, ydata, sigma=sigma, p0=param_p0, bounds=param_bounds, **kwargs
+    )
+    popt_err = np.sqrt(np.diag(pcov))
+
+    # Calculate the reduced chi-squared for fit
+    yfits = fit_func(xdata, *popt)
+    residues = (yfits - ydata) ** 2
+    if sigma is not None:
+        residues = residues / (sigma ** 2)
+    chisq = np.sum(residues) / (len(yfits) - num_params)
```

Comment:
Do we still want to name this field of the analysis result `chisq`, or change it to `reduced_chisq` ?  Also, maybe update the function documentation that reduced chi-square is calculated.

Reply:
We've had a discussion: if `xdata` consists of `n` points, and each was measured for `s` shots, should we fit `n` points, with the `n` means of `s` measurements, or should we fit `n*s` points? I argued that both are equivalent, because the objective function that's minimized by the least-squares algorithm is the same. This is still true (assuming that I was right in the first place), however after the fitting the returned reduced chi-square is different.

Reply:
For you second question: After discussing this a bit with @oliverdial, I think the answer was they are only equivalent under some assumptions of the underlying noise models being the same. Eg. sampling noise vs single-shot measurement noise, is it gaussian or not etc. In general the sigma values could be different in the two cases, P(0) for N-shots is binomial dist but sigma for each single shot depends on measurement discriminator error, so this can effect chi-squared and pcov